src/JestExt.ts

```diff
@@ -267,11 +274,7 @@ export class JestExt {
 
     this.coverageOverlay.enabled = updatedSettings.showCoverageOnLoad
 
-    this.stopProcess()
-
-    setTimeout(() => {

```

Comment:
I think the runner should report 'debuggerProcessExit event' in the normal situation (multiple times perhaps), but judging the fact that there is a timeout watch makes me wonder if there was incidence that it didn't... 

do you think it make sense to add the timer watch to `JestProcess` (not here) in case the stop promise didn't get resolved accordingly? If we are concering about the overhead, maybe add a extension option to control this behavior so users can turn it on when such rare situations occur...?

Reply:
The original message for 13921d74b05069c85bb9abdfb661b827af0f4046 doesn't read like it was specific to needing a timeout, so any (possible) dependencies on the timeout would probably be unintentional?

So are you thinking there should be a `setTimeout(resolve)` or a `Promise.race` in the `stop` method of `JestProcess`, as a fallback in case the stop hangs?

Reply:
hmmm... it seems to use a timer for rare edge case is not ideal... let's make sure the stopProcess will terminate the promise if it is still running (and I assume we will call stopProcess for every startProcess?). This way users can terminate the process themselves should it happen occasionally. what do you think?

Reply:
How would the force kill be accessed through the UI? Also, resolving the promise would only fix `stopAll` hanging, it wouldn't actually do any "killing" without additional logic.

Reply:
> How would the force kill be accessed through the UI

same as how we introduced new start/stop-all commands

>  resolving the promise would only fix stopAll hanging, it wouldn't actually do any "killing" without additional logic.

"fix hanging" is what "force kill" aim to do. Don't think we can do better in terms of "killing the process" after calling the `stop()`. All we try to avoid is a hanging promise preventing the subsequent start of new jestProcess.

However, I am having a second thought about the timeout, given `stop()` is only called during specific user interactions, such as users issue "Jest: Stop Runners" command, changing settings or update snapshots, i.e. it is not frequent. In that context, I think to have a timeout watching the zombie promise in `stop()` might be acceptable...

But all these might be moot points, depends on the underlying reason of the promise,  giving 1. we don't reuse the jestProcess in subsequent start() 2. there is nothing we attempt to do if it didn't exit as expected...  I can't help but wonder what the consequence for a  much simpler "kill-and-forget" pattern is? (i.e. the question I asked @marcinczenko above)

Reply:
let's try to close this thread, my 2 cents:
1. JestProcess can undoubtedly be cleaned up and streamlined, and hopefully, we can one day move to full Promise model (for the "start") instead of recursive invocation and complex state management... But it could be done in a separate PR.
1. regarding "stop", a promise is fine as we can still implement kill-and-forget in the timeout watch anyway. However, I struggled between a more forgiving approach (with an additional timeout to prevent zombie process blocking the subsequent start) and a more strict approach  (if the process didn't report close for any reason, no subsequent jest could be started). Considering the principle of _"do not block"_, I am leaning toward the more forgiving approach, which is also more compatible with the current behavior...

@vdh does that give you enough to proceed?

Reply:
@connectdotz So if I'm interpreting things correctly, we should avoid refactoring `JestProcess` for now in the scope of this PR, and just add some additional logic to prevent blocking? Would this change be suitable?

Remove the `await` inside `runTest` (revert my addition to that line) to avoid blocking there, and add a `Promise.race` to `stopProcess`?

```ts
export class JestExt {
  // …
  public async stopProcess() {
    this.channel.appendLine('Closing Jest')
    await Promise.race([
      this.jestProcessManager.stopAll(),
      new Promise((resolve) => { setTimeout(resolve, 500) }),
    ])
    this.status.stopped()
  }
  // …
}
```

I feel like the magic number on the timeout probably needs more refinement or labelling though.

The alternative would be to add a race inside `stopAll`, but then that would mean `stopJestProcess` would need changes as well, and that could get messy…?

Reply:
sorry for the late reply, can we encapsulate this stop logic inside the JestProcess itself?

Reply:
I should add this doesn't mean we need to refactor the rest of the JestProcess, I am thinking this can be easily encapsulated inside the JestProcess's `stop()` method without touching the rest.

Reply:
I've pushed an update with a `setTimeout` inside the `stop()` promise